An OrcaReader class.

muammar opened the pull request:
An OrcaReader class to operate together with https://github.com/OpenChemistry/mongochemdeploy/pull/72.

Commit:
An OrcaReader class.

This module can read orca output files and return a CJSON file to
operate with OpenChemistry.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+def OrcaReader(BaseReader):
+    """A class to parse orca output files and dump a chemical json file"""
+
+    def read(self, output_file):
```
The signature of this method does not conform with what the abstract base class requires: `def read(self)`.
The reader holds a reference to a file like object, which is passed to it in the constructor.
Which also means that any of the code below that uses `output_file` will need to be changed.

muammar replied:
I solved this.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+        molecule = next(readfile('orca', output_file))
+        obconv = openbabel.OBConversion()
+
+        # Generation of SMILES string
```
In general we have been saving inchi, inchikey, and smiles on the molecule model (which the calculation refers to). There is no need to add them again in the cjson attached to a calculation.

muammar replied:
That wasn't clear for me before. I removed those unneeded properties from the cjson object.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+                                         atom_numbers)))
+        molecular_mass = molecule.exactmass
+
+        cjson = {
```
The indentation of this dict got a little messed up when copying it over.

muammar replied:
According to `black` and `pycodestyle` checks, the structure of the dict is ok.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+                          }
+            }
+
+        energy = get_energy(output_file)
```
Does the cclib orca reader used below not extract the total energy? If it does, we shouldn't scan the whole output file twice.

muammar replied:
Yes, you were right. I went ahead and did it with `cclib`. I could not advance on this because there was a bug when reading these files. In fact, to test this PR one has to use the `master` repo until they release a new cclib version.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+            cjson.setdefault('vibrations', {})['frequencies'] = vibfreqs
+
+        if hasattr(data, 'vibdisps'):
+            vibdisps = _cclib_to_cjson_vibdisps(data.vibdisps)
```
`_cclib_to_cjson_vibdisps` is not being imported from anywhere

muammar replied:
Fixed.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+import openchemistry as oc
+from avogadro.core import Molecule
+from avogadro.io import FileFormatManager
+from jsonpath_rw import parse
```
Unused import

muammar replied:
All unused imports were now removed checking them using pyflakes.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+import json
+import openchemistry as oc
+from avogadro.core import Molecule
+from avogadro.io import FileFormatManager
```
unused import

muammar replied:
Done.

alesgenova commented on openchemistry/io/orca.py:
```diff
@@ -0,0 +1,121 @@
+        data = cclib.io.ccread(output_file)
+
+        if hasattr(data, 'gbasis'):
+            basis = oc.io.utils._cclib_to_cjson_basis(data.gbasis)
```
`_cclib_to_cjson_basis` is being imported from a local path, no need to specify `oc.io....`

muammar replied:
Fixed.

Commit:
Improvements based on review of io.orca

- After cclib's upstream parsing error was fixed for orca files it is
  possible to extract almost all needed information using that library.
- I removed the get_energy() function.
- I removed unneeded properties from the cjson object.
- IR vibrations are read correctly.
- The code is black clean, and pyflakes clean.

TODO:

Remove from read() the output_file argument.

Commit:
Remove argument out of read() method's signature.

Commit:
Added OrcaReader to __init__.py.

Commit:
Fixed OrcaReader class.

BaseReader was not inherited properly because of OrcaReader being
a function instead of a class. With this change, the signature of the
read() class method does conform with what the abstract base class
requires: def read(self).

muammar commented:
Hi @alesgenova. I made sure that using `docker run` to on the docker image effectively creates a CJSON file. When I try to test on jupyterlab I cannot use my personal image of orca;

![Screenshot from 2019-05-31 10-23-07](https://user-images.githubusercontent.com/954209/58723259-78fdbb80-838e-11e9-855b-c9ce8560af4b.png)

There is no error in the log widget. The orca docker image is located at https://cloud.docker.com/u/muammar/repository/docker/muammar/orca

 I think the image should work as the testing I did for the case of NWChem I only tested the `docker run` command worked. What do you think could be the problem?

Commit:
Merge branch 'master' into orca

* master: (21 commits)
  Make "3D coords generating" message less scary
  Check all molecules for 3D coords in batch job
  Unsure setuptools is up to date
  Add long_description_content_type='text/markdown'
  If 3D coords are not available, display SVG
  Automatically generate 3D coordinates if needed
  Added utilities to check if mol has 3d coords
  Update correct pending calculation before return
  For now, require all input parameters to be same
  If 3D coords are not available, display SVG
  Add an API function to run multiple calculations
  Move CalculationResult creation into function
  Allow _fetch_or_submit_calculation to run multiple
  Allow multiple calculations to run in one taskflow
  Update `GET /calculations` to new API
  Update GET /molecules/search API
  Update GET /molecules API to match new structure
  Make import_structure 3D coord generation optional
  Allow cjson to be `None` for GirderMolcule
  Expose `import_structure` to the public api
  ...

alesgenova reviewed (approved):
@muammar The reader class looks good to me. If you're running the container locally and it produces the CJSON you expect, I guess the reader is working fine. Which means the error you're seeing in the notebook must be related to the container?

muammar commented:
> @muammar The reader class looks good to me. If you're running the container locally and it produces the CJSON you expect, I guess the reader is working fine. Which means the error you're seeing in the notebook must be related to the container?

I think you are right about it. With recent updates, the logging widget is now showing progress and no error with the Orca output, but something else is failing at a later stage. Anyways, When I run the container locally it creates the files we are expected to work with OpenChemistry.
